File: 2021/07/01/main.go
package main

func grayCode(n int) []int {
	return gen([]int{0}, n, 1)
}

func gen(r []int, n, b int) []int {
	if n == 0 {
		return r
	}
	l := len(r)
	t := make([]int, l)
	for i := range r {
		t[l-i-1] = r[i] + b
	}

	return gen(append(r, t...), n-1, b*2)
}

File: 2021/07/01/main.py
class Solution:
    def grayCode(self, n: int) -> List[int]:
        def gen(r, n, b):
            if n == 0:
                return r
            l = len(r)
            t = [0] * l
            for i in range(l):
                t[l-i-1] = r[i] + b
            r += t
            return gen(r, n-1, b*2)
        return gen([0], n, 1)